File: Semester-3/Homework-1/LazyCalculation/LazyParallel.cs
using System;

namespace LazyCalculation
{
    /// <summary>
    /// Multi-thread lazy calculation class
    /// </summary>
    public class LazyParallel<T> : ILazy<T>
    {
        private Func<T> supplier;

        private volatile bool isCalculated = false;

        private T value;

        private object lockObject = new Object();

        public LazyParallel(Func<T> supplier)
        {
            if (supplier == null)
            {
                throw new ArgumentNullException();
            }
            this.supplier = supplier;
        }

        /// <summary>
        /// Get calculation result
        /// </summary>
        public T Get()
        {
            if (!isCalculated)
            {
                lock (lockObject)
                {
                    if (isCalculated)
                    {
                        return value;
                    }

                    value = supplier();
                    supplier = null;
                    isCalculated = true;
                }
            }
            
            return value;
        }
    }
}

File: Semester-4/Homework-7/Lazy/LazyParallel.fs
namespace Lazy

open System
open System.Threading

/// Multi thread version of Lazy
type LazyParallel<'a> (supplier : unit -> 'a) =
    let mutable result = None
    let lockObj = new Object()

    interface ILazy<'a> with
        
        /// Launches calculation and returns the result
        member this.Get () =
            match result with
            | None ->
                lock lockObj (fun () ->
                    match result with
                    | Some x -> x
                    | None ->
                        let value = supplier ()
                        result <- Some value
                        value
                    )
            | Some value ->
                value